torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2849,4 +2849,29 @@ void ProcessGroupGloo::monitoredBarrier(
             << " ms.";
 }
 
+void ProcessGroupGloo::setSequenceNumberForGroup() {
+  if (rank_ == 0) {
+    // Create and broadcast sequence number
+    auto seq = 1 + rand();
```

Comment:
Add a seed here to make the sequence number deterministic?

Reply:
Do we need determinism here? As I understand it, doesn't really matter what the # is, but we just need to increment it and keep it synchronized across processes. And for tests we just validate this behavior andn not the actual number?

Reply:
Can you explain why this initial seq number cannot be simply 0?

Reply:
Starting it at `0` might be tricky because there could be multiple process groups active within the application (maybe 1 NCCL and 1 Gloo pg). For keeping track of desynchronizations, this would still actually be fine. Although, for debug logs I'd prefer to start them off at different values so it's easy to keep track of which sequence of numbers belongs to which PG. What do you think?

Reply:
I think a random value in your implementation is totally fine.

Sometimes I just feel that it will be nice to know at which round a desync occurs. This can give the user a feeling how long does it take to hit the desync, e.g., at the very beginning? or took quite a while? or did different desync failures occur at around the same time?

Reply:
This is a good point, I think with the logging we will add with ProcessGroupWrapper, we will report:
1. sequence number
2. name and args of collective call

which hopefully helps make it a bit more clear when the desync happens. In addition, by logging the sequence no., we will be able to have a good idea as to how many collectives occured before the desync.

Reply:
> Starting it at `0` might be tricky because there could be multiple process groups active within the application (maybe 1 NCCL and 1 Gloo pg). For keeping track of desynchronizations, this would still actually be fine. Although, for debug logs I'd prefer to start them off at different values so it's easy to keep track of which sequence of numbers belongs to which PG. What do you think?
> 
> > how can we know at which round a desync occurs?
> 
> My plan is to increment the counter on each collective. Since to stay in sync every process should call the collectives the same no. of times in the same order, we will just communicate the sequence number (in detail debug mode) via the store and verify they are consistent across processes. This will help catch issues where a process "skips" a collective.

* NCCL kernel is launched asynchronously. CPU counter won't work to detect desync in most cases.
* Even to start from random seq, there is still the chance to conflict between different pgs. We can assign pg id to include the id in logs. By that mean we can just start form 0 and have good idea of how many collectives are called or finished.